src/Interpreters/join_common.cpp

```diff
@@ -164,6 +177,83 @@ Columns materializeColumns(const Block & block, const Names & names)
     return materialized;
 }
 
+/// Join keys not casted in some cases:
```

Comment:
```suggestion
/// We allow joins not only when the key type is the same on both sides, but 
/// also when the right side can be implicitly CAST'ed to the left type. To 
/// perform the join with implicit cast, we cast the right-side column to 
/// `Nullable(left type)`, and use this column as a join key. We substitute 
/// NULLs for non-convertible values. We then perform the join normally 
/// using this modified key column, which gets us the expected result for inner
/// and outer joins as well. We cast the right side because it is expected
/// to have less rows.
/// Join keys are not casted in some cases:
```

We need to have this high-level description of the algorithm somewhere. Not sure if I wrote it correctly, or where is the best place to put it.